### PR TITLE
dhcp6: rfc8415 fixes

### LINKED ISSUE #13460: networkd sends a DHCPv6 Information-Request on every received router advertisement
**systemd version the issue has been seen with**
current head, back at least as far as v237

**Used distribution**
Ubuntu

**Expected behaviour you didn't see**
When systemd-networkd receives a router advertisement with ND_RA_FLAG_MANAGED or ND_RA_FLAG_OTHER, it should enable the DHCPv6 client.

DHCPv6 is specified with its own timers for refreshing managed addresses and/or other configuration.  In particular, https://tools.ietf.org/html/rfc8415#section-21.23 dictates that the client must request an Information Refresh Time from the server when sending an Information-Request packet; systemd-networkd does not, and that should be implemented to comply with the RFC.

Regardless, in the absence of a server-specified Information Refresh Time, "the client MUST behave as if the option with the value IRT_DEFAULT was provided."  IRT_DEFAULT is defined as 86400 seconds (https://tools.ietf.org/html/rfc8415#section-7.6).

**Unexpected behaviour you saw**
systemd-network makes a DHCPv6 request *every* time it receives a router advertisement with the ND_RA_FLAG_OTHER flag set.

It is common for IPv6 routers to send router advertisements frequently -- generally at least every couple of minutes, but often every few seconds so that recovery from a failed router can be fast.  This causes systemd-networkd's DHCPv6 client to violate RFC 8415 by refreshing information more frequently than it should -- almost certainly more frequently than IRT_DEFAULT (1 day) and quite likely more frequently than IRT_MINIMUM (10 minutes).

With SYSTEMD_LOG_LEVEL=debug, the following messages are logged for *every* router advertisement:

```
Sep 03 14:38:29 hostname systemd-networkd[15730]: NDISC: Received Router Advertisement: flags OTHER preference medium lifetime 1800 sec
Sep 03 14:38:29 hostname systemd-networkd[15730]: NDISC: Invoking callback for 'r'.
Sep 03 14:38:29 hostname systemd-networkd[15730]: DHCPv6 CLIENT: Started in Information request mode
Sep 03 14:38:29 hostname systemd-networkd[15730]: bond0: Acquiring DHCPv6 lease on NDisc request
Sep 03 14:38:29 hostname systemd-networkd[15730]: DHCPv6 CLIENT: Sent INFORMATION-REQUEST
Sep 03 14:38:29 hostname systemd-networkd[15730]: DHCPv6 CLIENT: Next retransmission in 962ms
Sep 03 14:38:29 hostname systemd-networkd[15730]: bond0: Updating address: [elided]/64 (valid for 4w 2d)
Sep 03 14:38:29 hostname systemd-networkd[15730]: DHCPv6 CLIENT: Recv REPLY
```

This is a problem for us because we have VLANs containing a large number of systemd-networkd systems; on every router advertisement that we send, every one of those systems sends a DHCPv6 Information-Request packet within a 1ms window.  This swamps the server with a large number of synchronised requests and also swamps the network with extreme bursts of multicast traffic, triggering the switches' multicast storm protection etc..

**Steps to reproduce the problem**
On a network which uses DHCPv6 in stateless mode, observe how frequently a systemd system sends Information-Requests.

### PR DESCRIPTION
**Pull request by yuwata**
Closes #13460.

The first commit is not tested.

### CODE REVIEW

File: src/libsystemd-network/sd-dhcp6-client.c [hunk 1]
```diff
@@ -1025,6 +1035,8 @@ static int client_parse_message(
                 }
         }
 
+        client->information_refresh_time_usec = irt < IRT_MINIMUM ? IRT_MINIMUM : irt;
+
```
Comment (keszybz) [inline src/libsystemd-network/sd-dhcp6-client.c:1039]: MAX(IRT_MINIMIUM, irt) ?

File: src/libsystemd-network/sd-dhcp6-client.c [hunk 2]
```diff
@@ -1425,8 +1437,15 @@ int sd_dhcp6_client_start(sd_dhcp6_client *client) {
                 client->fd = r;
         }
 
-        if (client->information_request)
+        if (client->information_request) {
+                usec_t t = now(CLOCK_MONOTONIC);
+
+                if (t < client->information_request_time_usec + client->information_refresh_time_usec)
```
Comment (keszybz) [inline src/libsystemd-network/sd-dhcp6-client.c]: usec_add?

**Comment by yuwata** [comment]
@keszybz Thank you for the review. I've force-pushed a revised version. The above two points are addressed. PTAL.

File: src/libsystemd-network/sd-dhcp6-client.c [hunk 3]
```diff
@@ -29,6 +29,9 @@
 
 #define MAX_MAC_ADDR_LEN INFINIBAND_ALEN
 
+#define IRT_DEFAULT 1 * USEC_PER_DAY
+#define IRT_MINIMUM 600 * USEC_PER_SEC
```
Comment (keszybz) [inline src/libsystemd-network/sd-dhcp6-client.c:33]: It would be better to put both in parens.

**Comment by keszybz** [comment]
I think we should merge this, even though we don't have any specific test results confirming that the issue is fixed. If this is not doing what it should, somebody will notice ;)

**Pull request merged**